ystia/yorc
# Feature/split workflow exec into step

# Pull Request description
## Description of the change
This PR implements #93 

It allows to define a workflow step as unit execution task in order to split the workflow execution into different workers of one or many Yorc instances

### What I did
Define a task as an execution for non workflow task
Define a task as n executions for workflow task with n equal to the steps number
#### 1. Refactor the tasks Collector as to define Collector concerns:
- register in Consul the task (/tasks/)
- register in Consul the workflow if task is workflow (/workflows/)
- register in Consul the executions for initial steps if task is workflow (/executions/)
- register in Consul the execution for task if non workflow task
#### 2. Just change the dispatcher to handle task executions instead of tasks
Dispatcher concern is polling executions task and dispatch them across available workers
- It has to acquire a lock on the execution task as other distributed dispatchers can try to do the same
- If it gets the lock, it instantiates an execution task and push it to workers pool
- If it receives a message from shutdown channel, it has to spread the shutdown to workers
#### 3. Changes the worker concern to execute workflow step as execution instead of whole workflow
- worker executes a task execution sent from the dispatcher
- It has to poll cancellation and failure flags to stop execution
-  In case of workflow task, it runs a workflow step as a task execution and needs to update related task status, and register next step tasks execution if its execution is successful. In case of workflow join, it needs to check all previous steps are done.
- It has to release related execution Consul lock and mark the execution to delete
- If workflow is done, the task and deployment statuses need to be updated
- If an error occurred during the step execution, task status need to be updated
- Each done step will register the next ones. In case of join, the last done of previous steps will register the next step
- For a final step, it checks the status of all steps and if all are done, it sets the task status to done

### How to verify it
As this impacts many things in orchestrator, non-regression tests need to be run with all task types as custom command, custom workflow, query, deploy, undeploy, purge, cancel, resume in distributed and non-distributed modes.
To ensure, the goal is reached, a topology with 2 computes has to be deployed by a Yorc cluster with 2 nodes, each node with only one worker so that each node runs the install of one compute as these steps are parallel.


## Applicable Issues
#93
